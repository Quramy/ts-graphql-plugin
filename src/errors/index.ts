export { ErrorReporter } from './error-reporter';

export type ErrorRange = {
  fileName: string;
  start: number;
  end: number;
};

export type Severity = 'Error' | 'Warn';

export type ErrorContent = ErrorRange & {
  severity?: Severity;
  content: string;
};

export class ErrorWithLocation extends Error {
  readonly name = 'ErrorWithLocation';
  readonly severity: Severity = 'Error';

  constructor(public readonly message: string, public readonly errorContent: ErrorContent) {
    super(message);
    if (errorContent.severity) {
      this.severity = errorContent.severity;
    }
  }
}

export class ErrorWithoutLocation extends Error {
  readonly name = 'ErrorWithoutLocation';
  constructor(public readonly message: string, public readonly severity: Severity = 'Error') {
    super(message);
  }
}

export type TsGqlError = ErrorWithLocation | ErrorWithoutLocation;

export const ERROR_CODES = {
  graphqlLangServiceError: {
    code: 51001,
  },
  templateIsTooComplex: {
    code: 51010,
    message: 'This operation or fragment has too complex interpolation to analyze.',
  },
  errorInOtherInterpolation: {
    code: 51011,
    message: 'This expression has some GraphQL errors.',
  },
  schemaBuildError: {
    code: 51020,
  },
};
