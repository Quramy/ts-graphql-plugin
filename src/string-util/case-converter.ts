const STRING_DECAMELIZE_REGEXP = /([a-z\d])([A-Z])/g;
const STRING_DASHERIZE_REGEXP = /[ _]/g;

export function decamelize(str: string): string {
  return str.replace(STRING_DECAMELIZE_REGEXP, '$1_$2').toLowerCase();
}

export function dasherize(str: string): string {
  return decamelize(str).replace(STRING_DASHERIZE_REGEXP, '-');
}
