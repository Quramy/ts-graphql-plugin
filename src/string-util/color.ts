const resetCode = '\u001b[0m';
const colorCode = {
  thin: '\u001b[2m',
  invert: '\u001b[7m',
  black: '\u001b[30m',
  red: '\u001b[31m',
  green: '\u001b[32m',
  yellow: '\u001b[33m',
  blue: '\u001b[34m',
  magenta: '\u001b[35m',
  cyan: '\u001b[36m',
  white: '\u001b[37m',
};

export const color = Object.entries(colorCode).reduce((acc: any, [name, code]) => {
  return {
    ...acc,
    [name]: (msg: string) => code + msg + resetCode,
  };
}, {}) as {
  [P in keyof typeof colorCode]: (msg: string) => string;
};

export function clearColor(msg: string) {
  const outs: string[] = [];
  let i = 0;
  while (i < msg.length) {
    const charactor = msg[i++];
    if (charactor === '\u001b') {
      while (msg[i++] !== 'm') continue;
    } else {
      outs.push(charactor);
    }
  }
  return outs.join('');
}
