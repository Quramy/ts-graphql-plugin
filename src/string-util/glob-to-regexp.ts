export function globToRegExp(pattern: string) {
  let reStr = '';
  for (let i = 0; i < pattern.length; i++) {
    const char = pattern[i];

    switch (char) {
      case '/':
      case '$':
      case '^':
      case '+':
      case '.':
      case '(':
      case ')':
      case '=':
      case '!':
      case '|':
      case '[':
      case ']':
      case '{':
      case '}':
        reStr += '\\' + char;
        break;

      case '?':
        reStr += '.';
        break;

      case '*': {
        const prevChar = pattern[i - 1];
        let starCount = 1;
        while (pattern[i + 1] === '*') {
          starCount++;
          i++;
        }
        const nextChar = pattern[i + 1];
        const isGlobstar =
          starCount > 1 && (prevChar === '/' || prevChar == null) && (nextChar === '/' || nextChar == null);

        if (isGlobstar) {
          reStr += '((?:[^/]*(?:/|$))*)';
          i++;
        } else {
          reStr += '([^/]*)';
        }
        break;
      }

      default:
        reStr += char;
    }
  }

  reStr = '^' + reStr + '$';
  return new RegExp(reStr);
}
