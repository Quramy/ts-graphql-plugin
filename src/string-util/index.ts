export * from './padding';
export * from './case-converter';
export * from './color';
export * from './position-converter';
