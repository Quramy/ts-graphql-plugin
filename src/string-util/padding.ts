export function pad(letter: string, length: number) {
  const outs: string[] = [];
  for (let i = 0; i < length; i++) {
    outs.push(letter);
  }
  return outs.join('');
}
