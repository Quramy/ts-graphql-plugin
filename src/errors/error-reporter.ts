import path from 'path';
import { TsGqlError, ErrorWithLocation, ErrorWithoutLocation } from '.';
import { pos2location, pad, color } from '../string-util';

const lineMark = (line: number, width: number) => {
  const strLine = line + 1 + '';
  return color.invert(pad(' ', width - strLine.length) + strLine) + ' ';
};

const lineMarkForUnderline = (width: number) => {
  return color.invert(pad(' ', width)) + ' ';
};

export class ErrorReporter {
  constructor(private readonly _currentDirectory: string, private readonly _output: (msg: string) => void = () => {}) {}

  outputError(error: TsGqlError) {
    if (error instanceof ErrorWithoutLocation) {
      this._outputErrorWithoutLocation(error);
    } else if (error instanceof ErrorWithLocation) {
      this._outputErrorWithLocation(error);
    }
  }

  _outputErrorWithoutLocation(error: ErrorWithoutLocation) {
    const prefix = error.severity === 'Error' ? color.red('error') : color.yellow('warn');
    this._output(prefix + ': ' + error.message);
  }

  _outputErrorWithLocation(error: ErrorWithLocation) {
    const {
      message,
      errorContent: { content, start, end, fileName },
    } = error;
    const startLC = pos2location(content, start);
    const endLC = pos2location(content, end);
    const relativeContentPath = path.isAbsolute(fileName) ? path.relative(this._currentDirectory, fileName) : fileName;
    const fileIndicator = `${relativeContentPath}:${startLC.line + 1}:${startLC.character + 1}`;
    const outputs = [`${color.thin(fileIndicator)} - ${message}`, ''];
    const allLines = content.split('\n');
    const preLines = allLines.slice(Math.max(startLC.line - 1, 0), startLC.line);
    const lines = allLines.slice(startLC.line, endLC.line + 1);
    const postLines = allLines.slice(endLC.line + 1, Math.min(allLines.length - 1, endLC.line + 2));
    const lineMarkerWidth = (Math.min(allLines.length - 1, endLC.line + 2) + '').length;
    for (let i = 0; i < preLines.length; ++i) {
      outputs.push(lineMark(i + startLC.line - 1, lineMarkerWidth) + color.thin(preLines[i]));
    }
    for (let i = 0; i < lines.length; ++i) {
      outputs.push(lineMark(i + startLC.line, lineMarkerWidth) + lines[i]);
      if (i === 0) {
        if (startLC.line === endLC.line) {
          outputs.push(
            lineMarkForUnderline(lineMarkerWidth) +
              pad(' ', startLC.character) +
              color.red(pad('~', endLC.character - startLC.character)),
          );
        } else {
          outputs.push(
            lineMarkForUnderline(lineMarkerWidth) +
              pad(' ', startLC.character) +
              color.red(pad('~', lines[i].length - startLC.character)),
          );
        }
      } else if (i === lines.length - 1) {
        outputs.push(lineMarkForUnderline(lineMarkerWidth) + color.red(pad('~', endLC.character)));
      } else {
        outputs.push(lineMarkForUnderline(lineMarkerWidth) + color.red(pad('~', lines[i].length)));
      }
    }
    for (let i = 0; i < postLines.length; ++i) {
      outputs.push(lineMark(i + endLC.line + 1, lineMarkerWidth) + color.thin(postLines[i]));
    }
    outputs.push('');
    const result = outputs.join('\n');
    this._output(result);
    return result;
  }
}
