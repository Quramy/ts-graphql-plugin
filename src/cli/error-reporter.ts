import path from 'path';
import { ErrorWithLocation } from '../errors';
import { pos2location, pad, color } from '../string-util';

const lineMark = (line: number, width: number) => {
  const strLine = line + 1 + '';
  return color.invert(pad(' ', width - strLine.length) + strLine) + ' ';
};

const lineMarkForUnderline = (width: number) => {
  return color.invert(pad(' ', width)) + ' ';
};

export class ErrorReporter {
  constructor(private readonly _currentDirectory: string, private readonly _output: (msg: string) => void = () => {}) {}

  indicateErrorWithLocation(error: ErrorWithLocation) {
    const {
      message,
      errorContent: { content, start, end, fileName },
    } = error;
    const startLC = pos2location(content, start);
    const endLC = pos2location(content, end);
    const relativeContentPath = path.isAbsolute(fileName) ? path.relative(this._currentDirectory, fileName) : fileName;
    const fileIndicator = `${relativeContentPath}:${startLC.line + 1}:${startLC.character + 1}`;
    const outputs = [`${color.thin(fileIndicator)} - ${message}`, ''];
    const allLines = content.split('\n');
    const preLines = allLines.slice(Math.max(startLC.line - 1, 0), startLC.line);
    const lines = allLines.slice(startLC.line, endLC.line + 1);
    const postLines = allLines.slice(endLC.line + 1, Math.min(allLines.length - 1, endLC.line + 2));
    const lineMarkerWidth = (Math.min(allLines.length - 1, endLC.line + 2) + '').length;
    for (let i = 0; i < preLines.length; ++i) {
      outputs.push(lineMark(i + startLC.line - 1, lineMarkerWidth) + color.thin(preLines[i]));
    }
    for (let i = 0; i < lines.length; ++i) {
      outputs.push(lineMark(i + startLC.line, lineMarkerWidth) + lines[i]);
      if (i === 0) {
        if (startLC.line === endLC.line) {
          outputs.push(
            lineMarkForUnderline(lineMarkerWidth) +
              pad(' ', startLC.character) +
              color.red(pad('~', endLC.character - startLC.character)),
          );
        } else {
          outputs.push(
            lineMarkForUnderline(lineMarkerWidth) +
              pad(' ', startLC.character) +
              color.red(pad('~', lines[i].length - startLC.character)),
          );
        }
      } else if (i === lines.length - 1) {
        outputs.push(lineMarkForUnderline(lineMarkerWidth) + color.red(pad('~', endLC.character)));
      } else {
        outputs.push(lineMarkForUnderline(lineMarkerWidth) + color.red(pad('~', lines[i].length)));
      }
    }
    for (let i = 0; i < postLines.length; ++i) {
      outputs.push(lineMark(i + endLC.line + 1, lineMarkerWidth) + color.thin(postLines[i]));
    }
    outputs.push('');
    const result = outputs.join('\n');
    this._output(result);
    return result;
  }
}
