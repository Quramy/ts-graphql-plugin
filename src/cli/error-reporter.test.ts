import { ErrorReporter } from './error-reporter';
import { ErrorWithLocation, ErrorWithoutLocation } from '../errors';
import { mark, Markers } from '../string-util/testing/position-marker';
import { clearColor } from '../string-util';

describe(ErrorReporter, () => {
  describe(ErrorReporter.prototype.outputError, () => {
    describe(ErrorWithoutLocation, () => {
      it('should output error message', () => {
        let message: string = '';
        const reporter = new ErrorReporter('/prj', msg => (message = msg));
        reporter.outputError(new ErrorWithoutLocation('hoge', 'Error'));
        expect(clearColor(message)).toMatchSnapshot();
      });

      it('should output warn message', () => {
        let message: string = '';
        const reporter = new ErrorReporter('/prj', msg => (message = msg));
        reporter.outputError(new ErrorWithoutLocation('hoge', 'Warn'));
        expect(clearColor(message)).toMatchSnapshot();
      });
    });

    describe(ErrorWithLocation, () => {
      it('should output location of errors in human readable format', () => {
        let message: string = '';
        const reporter = new ErrorReporter('/prj', msg => (message = msg));
        const markers: Markers = {};
        reporter.outputError(
          new ErrorWithLocation('some error', {
            fileName: '/prj/main.ts',
            content: mark(
              `
          const query = invalidQuery;
          %%%           ^           ^  %%%
          %%%           a1          a2 %%%
        `,
              markers,
            ),
            start: markers.a1.pos,
            end: markers.a2.pos,
          }),
        );
        expect(clearColor(message)).toMatchSnapshot();
      });

      it('should output location of errors in human readable format with 2 lines', () => {
        let message: string = '';
        const reporter = new ErrorReporter('/prj', msg => (message = msg));
        const markers: Markers = {};
        reporter.outputError(
          new ErrorWithLocation('some error', {
            fileName: '/prj/main.ts',
            content: mark(
              `
          const query = gql\`;
            query MyQuery {
          %%%     ^             %%%
          %%%     a1            %%%
              name
          %%%     ^             %%%
          %%%     a2            %%%
            }
          \`:
        `,
              markers,
            ),
            start: markers.a1.pos,
            end: markers.a2.pos,
          }),
        );
        expect(clearColor(message)).toMatchSnapshot();
      });

      it('should output location of errors in human readable format with 3 or more lines', () => {
        let message: string = '';
        const reporter = new ErrorReporter('/prj', msg => (message = msg));
        const markers: Markers = {};
        reporter.outputError(
          new ErrorWithLocation('some error', {
            fileName: '/prj/main.ts',
            content: mark(
              `
          const query = gql\`;
            query MyQuery {
          %%%     ^             %%%
          %%%     a1            %%%
              id
              name
          %%%     ^             %%%
          %%%     a2            %%%
            }
          \`:
        `,
              markers,
            ),
            start: markers.a1.pos,
            end: markers.a2.pos,
          }),
        );
        expect(clearColor(message)).toMatchSnapshot();
      });
    });
  });
});
