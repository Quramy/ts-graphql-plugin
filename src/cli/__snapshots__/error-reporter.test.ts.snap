// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorReporter indicateErrorWithLocation should output location of errors in human readable format 1`] = `
"main.ts:2:25 - some error

1 
2           const query = invalidQuery;
                          ~~~~~~~~~~~~
"
`;

exports[`ErrorReporter indicateErrorWithLocation should output location of errors in human readable format with 2 lines 1`] = `
"main.ts:3:19 - some error

2           const query = gql\`;
3             query MyQuery {
                    ~~~~~~~~~
4               name
  ~~~~~~~~~~~~~~~~~~
5             }
"
`;

exports[`ErrorReporter indicateErrorWithLocation should output location of errors in human readable format with 3 or more lines 1`] = `
"main.ts:3:19 - some error

2           const query = gql\`;
3             query MyQuery {
                    ~~~~~~~~~
4               id
  ~~~~~~~~~~~~~~~~
5               name
  ~~~~~~~~~~~~~~~~~~
6             }
"
`;
